snewpdag/plugins/gen/TrueTimeDelay.py

```diff
@@ -0,0 +1,121 @@
+"""
+TrueTimeDelay: Calculate the true time delay of the neutrino arrival between two detectors
+by randomly generate a neutrino directional vector and a arrival time 
+
+Constructor Arguments:
+    detector_list: list of strings, ["first_detector", "second_detector", ...]
+                the list of detectors that we want to generate true time delay
+                options: "HK", "IC", "JUNO", "KM3", "SK"
+    detector_location: csv file ('detector_location.csv')
+
+Output added to data: dictionary with key ("det1", "det2") and the corresponding true time delay
+
+"""
+
+import csv
+import time
+import numpy as np
+import random as rm
+import itertools
+from datetime import datetime
+from snewpdag.dag import Node
+
+class TrueTimeDelay(Node):
+
+    #Define detector location
+    def __init__(self, detector_list, detector_location, **kwargs):
+        self.detector_info = {}
+        with open(detector_location, 'r') as f:
+            detectors = csv.reader(f)
+            for detector in detectors:
+                name = detector[0]
+                if name in detector_list: 
+                    lon = np.radians(float(detector[1]))
+                    lat = np.radians(float(detector[2]))
+                    height = float(detector[3])
+                    self.detector_info[name] = [lon, lat, height]
+        super().__init__(**kwargs)
+
+
+    #Randomly generate the direction vector for incoming neutrino flux, using right-ascention(alpha) and declination(delta)
+    def generate_n(self):
+        alpha_deg = rm.uniform(-180,180)
+        alpha = np.radians(alpha_deg)
+
+        delta_distribution = rm.uniform(-1,1)
+        delta = np.arccos(delta_distribution)
+        
+        nx = -np.cos(alpha)*np.cos(delta)
+        ny = -np.sin(alpha)*np.cos(delta)
+        nz = -np.sin(delta)
+        
+        source = np.array([nx, ny, nz])
+        return source
+
+
+    #randomly generate a time between 2000-1-1, 00:00 and 2000-12-31, 23:59:59 UTC for SN neutrino signal to arrive on Earth
+    #base time (vernal equinox): 2000-03-20, 12:00 PM UTC
+    def generate_time(self):
+        d = rm.randrange(0, 31536000)
+        start_range = datetime(2000,1,1,0,0,0)
+        start_unix = time.mktime(start_range.timetuple())
+        random_unix = d + start_unix 
+        generated_time = datetime.fromtimestamp(random_unix)
```

Comment:
You can keep the time in seconds, since later on you'll take the difference in seconds.

One also needs to keep track of fractions of a second.  In SNEWS we're likely to report times as two numbers:  something to give seconds in some defined epoch (like unix time), and then another integer counting nanoseconds.  For calculations within a plugin, it's sufficient to combine these into one number, seconds + nanoseconds*1e-9